CriticalMass/NetworkOperator.swift

```diff
@@ -18,36 +18,55 @@ struct NetworkOperator: NetworkLayer {
         self.networkIndicatorHelper = networkIndicatorHelper
     }
 
-    func get<T>(with url: URL, decodable: T.Type, completion: @escaping (T?) -> Void) where T: Decodable {
-        get(with: url, decodable: decodable, customDateFormatter: nil, completion: completion)
-    }
-
-    func get<T>(with url: URL, decodable: T.Type, customDateFormatter: DateFormatter?, completion: @escaping (T?) -> Void) where T: Decodable {
-        let request = URLRequest(url: url)
-        dataTask(with: request, decodable: decodable, customDateFormatter: customDateFormatter, completion: completion)
+    func get<T: APIRequestDefining>(request: T, completion: @escaping ResultCallback<T.ResponseDataType>) {
+        let urlRequest = request.makeRequest()
+        dataTask(with: urlRequest) { result in
+            switch result {
+            case let .failure(error):
+                completion(.failure(error))
+            case let .success(data):
+                do {
+                    let responseData = try request.parseResponse(data: data)
+                    completion(.success(responseData))
+                } catch {
+                    completion(.failure(NetworkError.parseError))
+                }
+            }
+        }
     }
 
-    func post<T>(with url: URL, decodable: T.Type, bodyData: Data, completion: @escaping (T?) -> Void) where T: Decodable {
-        var request = URLRequest(url: url)
-        request.httpMethod = "POST"
-        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
-        request.httpBody = bodyData
-        dataTask(with: request, decodable: decodable, customDateFormatter: nil, completion: completion)
+    func post<T: APIRequestDefining>(request: T, bodyData: Data, completion: @escaping ResultCallback<T.ResponseDataType>) {
+        var urlRequest = request.makeRequest()
+        urlRequest.httpBody = bodyData
+        dataTask(with: urlRequest) { result in
+            switch result {
```

Comment:
minor: there is some repetition in the result handling between post and get. I'm wondering if we could simplify that

Reply:
Good point. I was thinking of something like passing in the body as an optional and then only adding it if it has a value

Reply:
Oh. I confused your reply. I did not find a good (quick) solution. I want to find something tho and will park the ticket some time until I can work on a solution